### PR TITLE
RFC: time<->string conversions

### PR DESCRIPTION
**Pull request by knz**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9790)

### CODE REVIEW

File: docs/RFCS/time_conversion.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    - no working native Go implementation of strptime as of now,
+      so need to use a lib that wraps C's strptime with cgo.
+
+- Go Format/Parse:
+  - Pros: 
+    - performance: no call via cgo. 
+	- usability: format strings are more readable.
+  - Cons:
+    - mindshare: quite Go-specific.
+	- usability: timezone support is borken
```
Comment (mjibson) [inline docs/RFCS/time_conversion.md]: In what ways is it broken and worse than the strftime/strptime solution?

File: docs/RFCS/time_conversion.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	- usability: enables interleaving arbitrary text and numbers between formatted values.
+	- performance: parsing the format string is faster for the underlying function.
+  - Cons: 
+    - no working native Go implementation of strptime as of now,
+      so need to use a lib that wraps C's strptime with cgo.
+
+- Go Format/Parse:
+  - Pros: 
+    - performance: no call via cgo. 
+	- usability: format strings are more readable.
```
Comment (bdarnell) [inline docs/RFCS/time_conversion.md]: That's debatable. It looks nice for formats like

    RFC3339     = "2006-01-02T15:04:05Z07:00"

but a format string for the format used in our logs is pretty opaque: "060102 15:04:05". I find the `strptime` style more readable in this case: "%y%m%d %H:%M:%S"

**Review by bdarnell — COMMENTED** [review]
I vote for strptime/strftime because it's much more widely known.

**Comment by bdarnell** [comment]
In light of #10130, my preference for `strptime` still stands, but we should find or write (hopefully find) a self-contained `strptime` implementation so that the behavior doesn't vary based on the host platform.

**Comment by mjibson** [comment]
It seems like we should be able to write a quick package that converts the more common format strings to their Go equivalents. Would a simple string replacement map work? That would get us most of the strptime support (at least most of the stuff people actually want) with minimal effort and without any C dependencies.

**Comment by bdarnell** [comment]
We can get the most important parts pretty easily, but #10130 is about musl's lack of support for week numbers, which go doesn't support either. I think we need a complete posix-compatible implementation to do better than musl (although there's some value in saying `%U` is not supported by cockroachdb on any platform instead of having its support differ depending on the host platform).

**Comment by mjibson** [comment]
We can say that `%U` isn't supported by cockroachdb, but we should actually enforce that also. Just putting it in the docs isn't enough, since it still work on non-musl systems. I'm of the opinion that cockroachdb should work on heterogenous clusters and return identical results regardless of the machine type the client connected to. This draws me to the conclusion that we shouldn't be using any system's strptime functions since they differ. I still think doing a replacement over the go time package is the best solution for right now, because that will give us a framework to add the more esoteric posix flags.

Or is there some safe way to support the set of all strptime options, even if that means manually searching for `%U` and throwing an error or whatever in those cases? Are there any options for which linux/mac/musl return differing output for the same input (assuming we're only using options supported by all libs)?

**Comment by bdarnell** [comment]
A random data point: Python has its own portable implementation of strptime, which supports all the standard options. They fall back to the system implementation when any unknown percent directives are included, because these are apparently common enough that they can't just use their implementation all the time. But as long as they recognize all the directives, they never fall back to the system implementation because it might give different results, so it seems that there aren't any important cases where different results would matter within the standard subset.

**Comment by knz** [comment]
I like the python approach. 
Matt so you know there's already a Go package that tries to do what you say (google golang strptime) but they didn't do it well and that thing panics on invalid input. 
I'd rather embed a battle-tested implementation with a suitable license. There are multiple candidates to choose from really.

**Comment by bdarnell** [comment]
I'm not endorsing the python approach, just using it as evidence of the degree to which implementations vary. I think that for a SQL database, the argument for one consistent implementation everywhere is stronger than for allowing full access to the underlying platform's capabilities. So I'd embed some implementation without a fallback to the local platform.

**Comment by knz** [comment]
Flash news: C++11 standardized strftime/strptime via `std::strftime` and `std::time_get::get`.
Will probably finally close this issue by choosing that as common implementation.

**Comment by knz** [comment]
Yay to Bram and windows tests for revealing this!
Will help with #14383 and #10130

**Comment by tamird** [comment]
To throw another wrench into this plan, `tm.tm_gmtoff` and `tm.tm_zone` are BSD/GNU extensions, so they aren't present on windows. How does postgres parse time on Windows, I wonder?

**Comment by mjibson** [comment]
If all platforms had identical implementations then I'd begrudgingly be fine with using C to do this. However since that doesn't seem to be the case I think we should just write our own in Go. I'm planning on working on #3781 to address this, supporting the same syntax postgres uses.

**Comment by tamird** [comment]
Fantastic.

**Comment by mjibson** [comment]
Thinking through it more, I would argue to remove the experimental_strf/ptime functions before the 1.0 feature freeze. This would stop the C++ Windows problems this issue is causing and allow the more standard functions listed in that other issue to be implemented in a correct way.

**Comment by knz** [comment]
Wait hold your horses there.

1) the fact that some things do not *currently* work on Windows do not preclude us from providing the functionality on every other platform. Our "builtins" map allow this, also the prefix "experimental_" in the funbction name suggest we're making best effort at providing this feature.

2) as I found out the C++11 stdlib supports something like `strftime/`strptime`. This means we can rely on the functionality being available even on windows. If some fields are not standard, that just means we must abandon officially supporting / documenting the corresponding formatting codes (%XX) in cockroachdb. FWIW C++ indeed is silent on `tm_gmtoff` and `tm_zone` so I'd be fine making these silently partially supported. Again a little build tag fudgery is sufficient to support them where they are available.

3) if you indeed have a go at supporting `to_timestamp` and `from_timestamp` etc *a la* pg for compatibility, then I would *still* support a `strftime`/`strptime` interface *alongside* because the pg syntax is just horrendous and smells. (unless you decide to make the interface to the `to_`/`from_` functions accept the %-based formatting codes instead of pg's)

**Comment by mjibson** [comment]
These discussions are a good reason why we use Go instead of C.

Unless the implementation is exactly the same on all nodes, I think it is wrong to call the C functions. A heterogeneous cluster cannot return different results based on which gateway node was used. (This problem gets worse with DistSQL.)

**Comment by mjibson** [comment]
I have investigated using a non-C implementation of `strptime`. A Go implementation, done properly, is difficult. It is only medium difficult to get an implementation that supports the more common options by copying the code from the time package and making the changes to support the C-format strings (https://github.com/mjibson/cockroach/commit/6afe546a1d250ce169d8e2eaf254a5a9372909ba).

This supports most of the values that both implementations have in common. The z and Z options are not supported because the Go implementation pokes at unexported time data when parsing those. Copying over the time.Time type so we can poke those values also requires copying over all of the time.Location stuff which is where all the time zone loading comes in and it's pretty hairy.

The other big problem here is the strftime package we use doesn't have 1-1 matching for all strptime formats. strftime supports more formats (like %W). Also, if we ever did figure out the z options in strptime, strftime formats those differently, so strftime and strptime wouldn't always roundtrip. At that point, though, we'd just switch stfrtime to use our copied format implementation.

We *could* copy over pretty much the entire time package and change these strings we want. This would mean copying and maintaining all the system-dependent code for time zones (ugh). Also testing all of our new format strings. At this point we would have committed a substantial amount of resources for what is appearing to be a very expensive bike shed.

I think it is unfortunate that strftime and strptime have caused us so many problems. If they just worked the same everywhere then this wouldn't be a problem. But C doesn't have that guarantee. Go does.

My recommendation: retire these two functions away. Implement to_date and to_char using Go's time syntax. Reimplementing an entire time and zone parsing system is an incredible amount of work to do correctly. Go has already done it, even though some here don't like the synatx (I love it).

**Comment by knz** [comment]
I'd like to have a look at this again before we commit to a decision. Please leave the question open for a few more days until I come back.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by mjibson** [comment]
Yes, of course.

**Comment by bdarnell** [comment]
We discussed it this morning and propose to make a post on the forum asking if anyone is using these functions and add a deprecation warning to the release notes going out today. If no one speaks up, we can remove the `experimental_str?time` functions before 1.0.

**Comment by knz** [comment]
I have completed a platform-independent implementation at https://github.com/knz/strtime

Can we use this?

**Comment by knz** [comment]
see #14801

**Comment by knz** [comment]
Clarified the RFC to narrow down its scope to a mere comparison, leaving a decision to a later stage.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/time_conversion.md, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9790#-KTQL6oUnK0ioXEg2jTo-r1-70:-KkLeVZG5-gp84Wyi02A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ea312e02ae9e94a20a53fac0c8336927ce3c33/docs/RFCS/time_conversion.md#L70)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The cgo overhead shouldn't be a determining factor - we can always reimplement `strptime` in pure go if it turns out to be important for performance. 
</blockquote></details>

Done.

---

*[docs/RFCS/time_conversion.md, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9790#-KTQL6oUnK0ioXEg2jTo-r1-75:-KkLelPe_N-CWJyzF9f0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ea312e02ae9e94a20a53fac0c8336927ce3c33/docs/RFCS/time_conversion.md#L75)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

That's debatable. It looks nice for formats like

```
RFC3339     = "2006-01-02T15:04:05Z07:00"
```

but a format string for the format used in our logs is pretty opaque: "060102 15:04:05". I find the `strptime` style more readable in this case: "%y%m%d %H:%M:%S"

</blockquote></details>

Done.

---

*[docs/RFCS/time_conversion.md, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9790#-KTQL6oUnK0ioXEg2jTo-r1-78:-KkLgg_y8TRueKu9Ma2V:b-3lmf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5ea312e02ae9e94a20a53fac0c8336927ce3c33/docs/RFCS/time_conversion.md#L78)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

In what ways is it broken and worse than the strftime/strptime solution?
</blockquote></details>

I have clarified. The issue is minor really.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9790#-:-KkLgjUQ2id8Ut4HKjA8:b921haj)*

**Pull request merged**